feat: handle arrays in `fmt`

phaux opened the pull request:
Allow `fmt` to be called like a regular function with a single array argument. This was already kinda supported but:

1. The string parts array could only contain strings. Now it can contain stringables as well like the rest arguments.
2. You had to use both first arg array and rest args and their items would be zipped. After this change you can simply pass one array argument with mix of strings and stringables.
3. Nested arrays would be toStringed and formatting inside them would be lost. I changed it so nested arrays are handled properly.

KnightNiwrem commented:
> 3. Nested arrays would be toStringed and formatting inside them would be lost. I changed it so nested arrays are handled properly.

Thanks for the PQ! It's interesting, but I'm not so sure what is the intended goal of supporting (3). Could you present some cases for when (3) is needed?

phaux commented:
> Could you present some cases for when (3) is needed?

That was one of the first things that I tried to do after importing this library:

```js
ctx.replyFmt(
  fmt`Current queue:\n${tasks.map((task, idx) => fmt`${idx+1}. ${bold(task.user)}\n`)}`
)
```

And it didn't work. I guess (3) is not needed anymore thanks to (1) and (2), but that was actually the first thing I changed in my PR. Then I realised that just passing an array would be even nicer:

```js
ctx.replyFmt(fmt([
  "Current queue:\n",
  ...tasks.map((task, idx) => 
    fmt`${idx+1}. ${bold(task.user)}\n`
  ),
]))
```

So I implemented (1) and (2) too.

I guess we can revert that, but let's at least change the types so that TS says that array is not accepted. E.g. change `Stringable` to `FormattedString | string | number | null | undefined`. `Stringable` is just an interface with `toString` so it's basically equivalent to `Object` aka any type except nullish.

KnightNiwrem commented:
> > Could you present some cases for when (3) is needed?
> 
> That was one of the first things that I tried to do after importing this library:
> 
> ```js
> ctx.replyFmt(
>   fmt`Current queue:\n${tasks.map((task, idx) => fmt`${idx+1}. ${bold(task.user)}\n`)}`
> )
> ```
> 
> And it didn't work. I guess (3) is not needed anymore thanks to (1) and (2), but that was actually the first thing I changed in my PR. Then I realised that just passing an array would be even nicer:
> 
> ```js
> ctx.replyFmt(fmt([
>   "Current queue:\n",
>   ...tasks.map((task, idx) => 
>     fmt`${idx+1}. ${bold(task.user)}\n`
>   ),
> ]))
> ```
> 
> So I implemented (1) and (2) too.
> 
> I guess we can revert that, but let's at least change the types so that TS says that array is not accepted. E.g. change `Stringable` to `FormattedString | string | number | null | undefined`. `Stringable` is just an interface with `toString` so it's basically equivalent to `Object` aka any type except nullish.
> 
> Also supporting arrays brings it closer to jsx which _almost_ lets us do:
> 
> ```js
> /** @jsx fmt */
> ctx.replyFmt(
>   <>
>     Current queue:<br/>
>     {tasks.map((task, idx) =>
>       <>
>         {idx+1}. <b>{task.user}</b><br/>
>       </>
>     )}
>   </>
> )
> ```
> 
> Maybe in a future PR :3

I'm satisfied with (1) and (2) for now. Changing `Stringable` to the proposed union type is 1) a breaking change, and 2) in conflict with the intended design goal of making it `Stringable` in the first place. Hence, that proposal is rejected.

Allowing any object that has `toString()` implemented is by design, not a bug.

phaux commented:
@KnightNiwrem reverted (3)

phaux commented:
@KnightNiwrem @KnorpelSenf anything else I can do?

KnightNiwrem commented:
> @KnightNiwrem @KnorpelSenf anything else I can do?

Thanks for the PQ!
